addon/-private/class/modifier.ts

```diff
@@ -0,0 +1,126 @@
+ * update if any of those values change.
+ */
+export default class ClassBasedModifier<
+  Args extends ModifierArgs = ModifierArgs
+> {
+  // NOTE: these are private (and `DESTROYING` and `DESTROYED` are not exported
+  // in the main export, only at `-private` locations), but these cannot be made
+  // `private` in the class body because the modifier manager needs to be able
+  // to access them
+  [DESTROYING] = false;
+  [DESTROYED] = false;
+
+  /**
+   * The arguments passed to the modifier. `args.positional` is an array of
+   * positional arguments, and `args.named` is an object containing the named
+   * arguments.
+   */
+  readonly args: Args;
+
+  /**
+   * The element the modifier is applied to.
+   *
+   * @warning `element` is ***not*** available during `constructor` or
+   *   `willDestroy`.
+   */
+  element: Element = null as any;
+
+  constructor(owner: unknown, args: Args) {
+    setOwner(this, owner);
+    this.args = args;
+  }
+
+  /**
+   * Called when the modifier is installed **and** anytime the arguments are
+   * updated.
+   */
+  didReceiveArguments() {}
+
+  /**
+   * Called anytime the arguments are updated but **not** on the initial
+   * install. Called before `didReceiveArguments`.
+   */
+  didUpdateArguments() {}
+
+  /**
+   * Called when the modifier is installed on the DOM element. Called after
+   * `didReceiveArguments`.
+   */
+  didInstall() {}
+
+  /**
+   * Called when the DOM element is about to be destroyed; use for removing
+   * event listeners on the element and other similar clean-up tasks.
+   */
+  willRemove() {}
+
+  /**
+   * Called when the modifier itself is about to be destroyed; use for teardown
+   * code. Called after `willRemove`. The element is no longer available at this
+   * point (i.e. its value is `null` during teardown).
+   *
+   * @note TypeScript users can import the `InTeardown` type utility for safely
+   *   handling `null` here; see [the note in the README for details][README].
+   *
+   * [README]: https://github.com/ember-modifier/ember-modifier#lifecycle-hooks-and-types
+   */
+  willDestroy() {}
+
+  get isDestroying() {
+    return this[DESTROYING];
+  }
+
+  get isDestroyed() {
+    return this[DESTROYED];
+  }
+}
+
+/**
+ * A utility for extra type safety in the `willDestroy` lifecycle hook in
+ * classes which extend `ClassBasedModifier`: it types `element` as `null`.
+ *
+ * ## Usage
+ *
+ * ```ts
+ * import Modifier, { InTeardown } from 'ember-modifier';
+ *
+ * function doSomethingWithElement(el: Element) {
+ *   // ...
+ * }
+ *
+ * export default class CorrectlyTypedModifier extends Modifier {
+ *   willRemove(this: InTeardown<CorrectlyTypedModifier>) {
+ *     doSomethingWithElement(this.element); // TYPE ERROR!
+ *   }
+ * }
+ * ```
+ */
+export type InTeardown<M extends ClassBasedModifier> = Omit<M, "element"> & {
+  element: null;
+};
+
+setModifierManager(() => Manager, ClassBasedModifier);
```

Comment:
Any reason not to do this all in the export?

```ts
export default setModifierManager(() => Manager, class Modifier {

});
```

Reply:
Seems reasonable to me. I'll flag it up for one of the follow-up cleanups.

Reply:
Actually, on further consideration: yeah, we do actually have a good reason to define it independently: consumers need to be able to import the class to subclass!

Reply:
Hmm, that doesn't get changed in my example. `setModifierManager` returns the second arg, so `setModifierManager(someFunc, Foo)` results in `Foo` (and I think the types do show that, but I could be wrong).

Reply:
Interesting; I misunderstood the RFC—I thought (and the types I did for this repo, since there are none public at present) that it returned `void`. I’ll update the local type definitions to match that, so that when we upstream they’re correct.

The other consideration, though, is the `InTeardown` utility/hack type. Since there would be no named type for it to refer to for defining itself, there would be no way to make it work. Given that, I think I’d suggest we keep it as is for now, and then switch it to that form if/when we make a breaking release with some of the other changes we’ve gathered here?

Reply:
Yep, definitely seems good